beacon-chain/blockchain/execution_engine.go

```diff
@@ -250,22 +253,23 @@ func (s *Service) notifyNewPayload(ctx context.Context, postStateVersion int,
 
 // getPayloadAttributes returns the payload attributes for the given state and slot.
 // The attribute is required to initiate a payload build process in the context of an `engine_forkchoiceUpdated` call.
-func (s *Service) getPayloadAttribute(ctx context.Context, st state.BeaconState, slot types.Slot) (bool, *enginev1.PayloadAttributes, types.ValidatorIndex, error) {
+func (s *Service) getPayloadAttribute(ctx context.Context, st state.BeaconState, slot types.Slot) (bool, payloadattribute.Attributer, types.ValidatorIndex, error) {
+	emptyAttri := payloadattribute.EmptyWithVersion(st.Version())
 	// Root is `[32]byte{}` since we are retrieving proposer ID of a given slot. During insertion at assignment the root was not known.
 	proposerID, _, ok := s.cfg.ProposerSlotIndexCache.GetProposerPayloadIDs(slot, [32]byte{} /* root */)
 	if !ok { // There's no need to build attribute if there is no proposer for slot.
-		return false, nil, 0, nil
+		return false, emptyAttri, 0, nil
 	}
 
 	// Get previous randao.
 	st = st.Copy()
 	st, err := transition.ProcessSlotsIfPossible(ctx, st, slot)
 	if err != nil {
-		return false, nil, 0, err
+		return false, emptyAttri, 0, err
```

Comment:
Do we want to call FCU if `getPayloadAttribute` fails? If the answer is yes, then we must know what version the attribute is. Returning `emptyAttri` instead of `nil` embeds the version info for subsequent calls

```go
if attrs == nil {
		return nil, nil, errors.New("nil payload attribute")
	}
	switch attrs.Version() {
	case version.Bellatrix:
		a, err := attrs.PbV1()
		if err != nil {
			return nil, nil, err
		}
		err = s.rpcClient.CallContext(ctx, result, ForkchoiceUpdatedMethod, state, a)
		if err != nil {
			return nil, nil, handleRPCError(err)
		}
	case version.Capella:
		a, err := attrs.PbV2()
		if err != nil {
			return nil, nil, err
		}
		err = s.rpcClient.CallContext(ctx, result, ForkchoiceUpdatedMethodV2, state, a)
		if err != nil {
			return nil, nil, handleRPCError(err)
		}
	}
```

Reply:
Yes, but then we don't want to return `err` in all those calls and return the `emtpyAttri` with `nil` for `err`. And if we want to error out instead of calling `FCU` then we need to return `nil` instead of `emptyAttri` and the  `err` for the error. So I think in all the places that I remarked, we should be doing one of the two above, and not the combination `emptyAttri, err`. I think the most robust thing to do is to never return an error, change the signature of this function and call FCU all the times. But we need to check if there is an honest error condition that makes this impossible,